core/src/net/tcp_stream.cpp

```diff
@@ -0,0 +1,183 @@
+#ifdef _WIN32
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#pragma comment(lib, "Ws2_32.lib")
+#else
+#include <arpa/inet.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#endif
+
+#include <cstring>
+
+#include <cubos/core/log.hpp>
+#include <cubos/core/net/tcp_stream.hpp>
+#include <cubos/core/reflection/external/cstring.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
+#include <cubos/core/reflection/external/string.hpp>
+
+using cubos::core::net::Address;
+using cubos::core::net::TcpStream;
+
+/// @brief Cross-platform utility to get last error message.
+/// @return Message.
+/// @todo Move this to a higher module for general purposes?
+static std::string error()
+{
+#ifdef _WIN32
+    static char message[256] = {0};
+    FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, 0, WSAGetLastError(), 0, message, 256, 0);
+    char* nl = strrchr(message, '\n');
+    if (nl)
+        *nl = 0;
+    return {message};
+#else
+    return {std::strerror(errno)};
+#endif
+}
+
+bool TcpStream::connect(const Address& address, uint16_t port)
+{
+    mSock = socket(AF_INET, SOCK_STREAM, 0);
+    if (mSock == -1)
+    {
+        CUBOS_ERROR("Failed to allocate TCP socket: {}", error());
+        return false;
+    }
+
+    sockaddr_in addr;
+    addr.sin_family = AF_INET;
+    addr.sin_addr.s_addr = inet_addr(address.toString().c_str());
+    addr.sin_port = htons(port);
+
+    if (::connect(mSock, (struct sockaddr*)&addr, sizeof(addr)) < 0)
+    {
+        CUBOS_ERROR("Failed to connect TCP stream to {}:{}: {}", address.toString(), port, error());
+        return false;
+    }
+
+    CUBOS_INFO("New TCP stream at address {} and port '{}'", address.toString(), port);
+    return true;
+}
+
+bool TcpStream::connect(SOCKET socket)
+{
+    mSock = socket;
+    CUBOS_INFO("New TCP stream at socket {}", mSock);
+    return mSock != INVALID_SOCKET;
+}
```

Comment:
As I said above, I would prefer this to be renamed to `inner(SOCKET socket)`, as it is a setter.
Also, don't forged to close the old `mSock`, if it wasn't -1.